File: stress.py
import numpy as np
import pandas as pd
import networkx as nx
import matplotlib.pyplot as plt
from sklearn.metrics import pairwise_distances
from scipy.sparse.linalg import spsolve
import matplotlib.ticker as ticker
from sklearn.neighbors import NearestNeighbors
import time
from random import random
from networkx import floyd_warshall_numpy
import numpy as np
from scipy.sparse import csr_matrix
from scipy.sparse.linalg import cg
# プロット用のフォント設定
plt.rcParams["font.family"] = "DejaVu Serif"
plt.rcParams["font.size"] = 20


def inv(x):
    if x < 1e-5:
        return 0.0
    return 1 / x


def stress(X, D):
    """
    X: (n, 1) 配列、各ノードの1次元座標
    D: (n, n) ターゲット距離行列
    """
    n = len(X)
    s = 0.0
    for i in range(n):
        x_i = X[i, 0]  # 1次元なのでスカラー
        for j in range(i):
            x_j = X[j, 0]
            d = abs(x_i - x_j) - D[i, j]
            s += d * d
    return s


def stress_majorization(graph):
    epsilon = 1e-10
    n = graph.number_of_nodes()

    # ノード間の最短経路距離行列 D を算出
    D = floyd_warshall_numpy(graph)

    # 重み行列 w, および補正項 delta を計算
    w = np.zeros((n, n))
    delta = np.zeros((n, n))
    for i in range(n):
        for j in range(i):
            w[i, j] = w[j, i] = D[i, j] ** -2
            delta[i, j] = delta[j, i] = w[i, j] * D[i, j]

    # 初期配置 Z (1次元配列: shape (n, 1))
    Z = np.random.rand(n, 1)
    # ノード0を (0,) に固定（平行移動不変性の除去）
    Z[0, 0] = 0.0

    # 重み付きラプラシアン L_w の計算
    L_w = np.zeros((n, n))
    for i in range(n):
        for j in range(i):
            L_w[i, j] = L_w[j, i] = -w[i, j]
    for i in range(n):
        L_w[i, i] = -sum(L_w[i, :])

    # 初期ストレスの評価
    e0 = stress(Z, D)

    # 反復更新
    while True:
        # 補助ラプラシアン L_Z の構築
        L_Z = np.zeros((n, n))
        for i in range(n):
            for j in range(i):
                # 1次元の場合、Z[i]-Z[j] はスカラーなので abs() で距離を計算
                L_Z[i, j] = L_Z[j, i] = -delta[i, j] * inv(abs(Z[i, 0] - Z[j, 0]))
        for i in range(n):
            L_Z[i, i] = -sum(L_Z[i, :])

        # 更新：ノード0は固定なので、ノード1以降の座標のみ更新
        # 連立方程式 L_w_reduced * x_new = (L_Z @ Z) から解く
        rhs = (L_Z @ Z[:, 0])[1:]
        x_new_reduced, info = cg(L_w[1:, 1:], rhs, atol=epsilon)
        if info != 0:
            print(f"CG did not converge, info = {info}")
        Z[1:, 0] = x_new_reduced

        # 新たなストレスを計算して収束判定
        e = stress(Z, D)
        if (e0 - e) / e0 < epsilon:
            break
        e0 = e

    return Z

def stress_majorization_1d(dist_matrix, max_iter=1000, tol=1e-4):
    """
    全ペアを使用する従来型の1次元Stress Majorization（比較用）
    """
    N = dist_matrix.shape[0]
    epsilon = 1e-20
    dist_matrix = np.maximum(dist_matrix, epsilon)
    W = 1.0 / dist_matrix ** 1
    #W = np.ones(dist_matrix.shape)
    Lw = np.diag(W.sum(axis=1)) - W
    x = np.random.rand(N)
    x = x - np.mean(x)
    stress_prev = np.inf
    for iteration in range(max_iter):
        delta = W * dist_matrix
        diff = np.subtract.outer(x, x)
        norm_val = np.maximum(np.abs(diff), epsilon)
        B = delta * diff / norm_val
        row_sum = B.sum(axis=1)
        Lz = np.diag(row_sum) - B
        #x_new = spsolve(csr_matrix(Lw), Lz @ x)
        x_new, info = cg(csr_matrix(Lw), Lz @ x, atol=tol)
        x_new = x_new - np.mean(x_new)
        norm_diff = np.abs(np.subtract.outer(x_new, x_new))
        stress = np.sum(W * (norm_diff - dist_matrix) ** 2) / 2
        if (stress_prev - stress) / stress_prev < tol:
            print(f"Converged after {iteration + 1} iterations, stress = {stress:.4e}")
            return x_new
        x = x_new
        stress_prev = stress
    print(f"Max iterations reached, final stress = {stress:.4e}")
    return x_new

start = time.time()
# --- データ読み込みと前処理 ---
problem_name = 'RWMOP6'
algo = 'data'
domain_df = pd.read_csv('domain_info.csv')
row = domain_df.loc[domain_df['problem'] == problem_name].iloc[0]
lower = np.array([float(v) for v in row['lower'].split(",")])
upper = np.array([float(v) for v in row['upper'].split(",")])
diff = upper - lower
data = pd.read_csv(f'data09-20/{problem_name}_{algo}.csv')
#data = pd.read_csv(f'data09-20-pre/local_search{problem_name}.csv')

# 各制約違反値の計算（複数のCon_がある場合は0以上の和をとる）
con_cols = [c for c in data.columns if c.startswith('Con_')]
total = data[con_cols].apply(lambda row: np.sum(np.maximum(0, row)), axis=1)
data['CV'] = total
data_sorted = data.sort_values(by=['ID', 'Gen'])

# --- グラフ構築 ---
G = nx.DiGraph()
X_cols = [c for c in data.columns if c.startswith('X_')]
for idx, row in data_sorted.iterrows():
    G.add_node(idx, Gen=row['Gen'], ID=row['ID'], X=row[X_cols].values, CV=row['CV'])
prev_row = None
prev_idx = None
for idx, row in data_sorted.iterrows():
    if prev_row is not None and prev_row['ID'] == row['ID'] and row['Gen'] == prev_row['Gen'] + 1:
        G.add_edge(prev_idx, idx)
    prev_row = row
    prev_idx = idx

# --- 重複ノードの統合 ---
from collections import defaultdict

vec2nodes = defaultdict(list)
for n in G.nodes():
    vec = tuple(G.nodes[n]['X'])
    vec2nodes[vec].append(n)
for vec, node_list in vec2nodes.items():
    if len(node_list) > 1:
        representative = node_list[0]
        for dup in node_list[1:]:
            for pred in list(G.predecessors(dup)):
                if pred != representative and not G.has_edge(pred, representative):
                    G.add_edge(pred, representative)
            for succ in list(G.successors(dup)):
                if succ != representative and not G.has_edge(representative, succ):
                    G.add_edge(representative, succ)
            G.remove_node(dup)

nodes = list(G.nodes())
a = nx.number_of_nodes(G)
print(a)

X_all = np.array([G.nodes[n]['X'] for n in nodes])
X_all_norm = (X_all - lower) / diff
# --- 設計変数に基づくユークリッド距離行列 ---
dist_matrix = pairwise_distances(X_all_norm, metric='euclidean')
#x_coords = stress_majorization_1d(dist_matrix)
x_coords = stress_majorization(G)

# --- 可視化 ---
pos = {n: (x_coords[i], G.nodes[n]['CV']) for i, n in enumerate(nodes)}

# ノードの分類
sink_nodes = [n for n in G.nodes() if G.out_degree(n) == 0]
final_feasible = [n for n in sink_nodes if G.nodes[n]['CV'] == 0]
final_infeasible = [n for n in sink_nodes if G.nodes[n]['CV'] > 0]
other_nodes = [n for n in nodes if n not in sink_nodes]
midle_feasible = [n for n in other_nodes if G.nodes[n]['CV'] == 0]
midle_infeasible = [n for n in other_nodes if G.nodes[n]['CV'] > 0]

plt.figure(figsize=(15, 12))
nx.draw_networkx_edges(G, pos, arrowstyle='->', arrowsize=15, edge_color='gray', alpha=0.5)
nx.draw_networkx_nodes(G, pos, nodelist=midle_infeasible, node_size=50, node_color='skyblue', edgecolors='black')
nx.draw_networkx_nodes(G, pos, nodelist=midle_feasible, node_size=50, node_color='salmon', edgecolors='black')
nx.draw_networkx_nodes(G, pos, nodelist=final_feasible, node_size=100, node_color='red', edgecolors='black')
nx.draw_networkx_nodes(G, pos, nodelist=final_infeasible, node_size=100, node_color='blue', edgecolors='black',
                       node_shape='o', linewidths=2)

ax = plt.gca()
ax.set_yscale('symlog', linthresh=1e-6)
log_formatter = ticker.LogFormatterSciNotation(base=10)
ax.yaxis.set_major_formatter(log_formatter)
ax.tick_params(axis='y', which='both', labelleft=True)
ax.axis('on')
plt.ylim(bottom=-1e-6, top=1e5)
plt.show()
end = time.time()  # 現在時刻（処理完了後）を取得
time_diff = end - start  # 処理完了後の時刻から処理開始前の時刻を減算する
print(time_diff)  # 処理にかかった時間データを使用

File: stress.R
# ---------------------------------------------------------
# 必要なライブラリをロード
# ---------------------------------------------------------
library(dplyr)
library(tidyr)
library(igraph)      # グラフ作成・操作用
library(ggplot2)     # 可視化用
library(scales)      # ログスケール表示など
library(Matrix)
library(Rlinsolve)

# ---------------------------------------------------------
# 1. データ読み込みと前処理
# ---------------------------------------------------------
problem_name <- "RWMOP22"

# domain_info.csv を読み込んで、その中から該当問題行を取得
domain_df <- read.csv("/Users/azumayuki/Documents/LONs/domain_info.csv", stringsAsFactors=FALSE)
row_domain <- domain_df %>% filter(problem == problem_name) %>% slice(1)

# lower, upper を数値ベクトル化
lower <- as.numeric(unlist(strsplit(row_domain$lower, ",")))
upper <- as.numeric(unlist(strsplit(row_domain$upper, ",")))
diff <- upper - lower

# ローカルサーチ結果などが入った CSV を読み込む
# ファイルパスは実際の環境に合わせて修正してください。
data_raw <- read.csv(file.path("/Users/azumayuki/Documents/LONs/data09-20-pre", paste0("local_search", problem_name, ".csv")),
                     stringsAsFactors=FALSE)

# 制約列 (Con_*) を探し合計CVを計算
con_cols <- grep("^Con_", colnames(data_raw), value=TRUE)
data_raw <- data_raw %>%
  rowwise() %>%
  mutate(CV = sum(pmax(0, c_across(all_of(con_cols))))) %>%
  ungroup()

# 世代 (Gen) とID (ID) で並べ替え
data_sorted <- data_raw %>% arrange(ID, Gen)

# ---------------------------------------------------------
# 2. グラフ作成 (igraph) 
# ---------------------------------------------------------
# まずノード情報を作る
#   項目: Gen, ID, X(ベクトル), CV
#   ノードの名前は行インデックスにしておく
# ---------------------------------------------------------
# X列をまとめる
x_cols <- grep("^X_", colnames(data_sorted), value=TRUE)

nodes_df <- data_sorted %>%
  mutate(idx = row_number()) %>%
  select(idx, ID, Gen, CV, all_of(x_cols)) %>%
  as.data.frame()  # <- as.data.frame()を挟む

rownames(nodes_df) <- nodes_df$idx

# ---------------------------------------------------------
# 3. エッジの作成
#    同一ID かつ Genが連続 (g+1=次世代) の場合にエッジを張る
# ---------------------------------------------------------
edge_list <- list()
prev_row <- NULL

for (i in seq_len(nrow(nodes_df))) {
  current_row <- nodes_df[i, ]
  if (!is.null(prev_row)) {
    # 同じIDで, 世代が1つ上がっていたらエッジ
    if (prev_row$ID == current_row$ID && 
        current_row$Gen == prev_row$Gen + 1) {
      edge_list <- append(edge_list, list(c(prev_row$idx, current_row$idx)))
    }
  }
  prev_row <- current_row
}

edges_df <- do.call(rbind, edge_list)
colnames(edges_df) <- c("from", "to")
edges_df <- as.data.frame(edges_df)

# いったんグラフ構築
g <- graph_from_data_frame(d = edges_df, 
                           vertices = nodes_df,
                           directed = TRUE)

# ---------------------------------------------------------
# 4. 重複ノードの統合 (まったく同じ X ベクトルを持つノード)
# ---------------------------------------------------------
#   - 同じ設計変数 X のノードを1つの代表ノードにまとめる
#   - 代表以外のノードに入っていたエッジは代表にリダイレクトし、ノードを削除
# ---------------------------------------------------------

# X を文字列化して重複判定
paste_x_values <- function(rowX) {
  paste(format(rowX, digits=15), collapse=",")
}
print(x_cols)
# 各ノードの X ベクトルを文字列に
x_strings <- apply(as.matrix(nodes_df[x_cols]), 1, paste_x_values)

# 同じ文字列を持つノード達をグループ化
x_groups <- split(names(x_strings), f=x_strings)  # names(...) は rownames(nodes_df) = idx

# 重複ノードを1つの代表にまとめる
for (group_key in names(x_groups)) {
  group_nodes <- x_groups[[group_key]]
  if (length(group_nodes) > 1) {
    # 代表ノード(先頭)を決める
    rep_node <- group_nodes[1]
    dup_nodes <- group_nodes[-1]
    # 重複ノードの predecessor を代表にリダイレクト
    for (dn in dup_nodes) {
      preds <- igraph::neighbors(g, dn, mode="in")
      for (p in preds) {
        if (p$name != rep_node && 
            !are.connected(g, p$name, rep_node)) {
          g <- add_edges(g, c(p$name, rep_node))
        }
      }
      # 重複ノードの successor を代表にリダイレクト
      succs <- igraph::neighbors(g, dn, mode="out")
      for (s in succs) {
        if (s$name != rep_node && 
            !are.connected(g, rep_node, s$name)) {
          g <- add_edges(g, c(rep_node, s$name))
        }
      }
      # 重複ノード削除
      g <- delete_vertices(g, dn)
    }
  }
}

stress_majorization_1d <- function(dist_matrix, max_iter = 1000, tol = 1e-4) {
  N <- nrow(dist_matrix)
  epsilon <- 1e-10
  
  # 0距離を回避
  dist_matrix[dist_matrix < epsilon] <- epsilon
  
  # W = 1 / dist_matrix
  W <- 1 / dist_matrix
  
  # Lw = diag(rowSums(W)) - W
  rowSumsW <- rowSums(W)
  Lw <- diag(rowSumsW) - W
  
  # 疎行列化 (共役勾配法のA行列)
  Lw_sp <- Matrix(Lw, sparse=TRUE)
  
  # 初期解
  x <- runif(N)
  x <- x - mean(x)
  
  stress_prev <- Inf
  
  for (iter in seq_len(max_iter)) {
    # δ = W * dist_matrix
    delta <- W * dist_matrix
    
    # diff[i,j] = x[i] - x[j]
    diff <- outer(x, x, FUN = "-")
    norm_val <- pmax(abs(diff), epsilon)
    
    # B = delta * diff / norm_val
    B <- delta * diff / norm_val
    row_sum_B <- rowSums(B)
    Lz <- diag(row_sum_B) - B
    
    # 右辺
    rhs <- Lz %*% x
    init_vec <- rep(Inf, length(rhs))
    # ---- CGで解く ----
    sol <- lsolve.cg(A = Lw_sp, B = rhs, xinit = init_vec, reltol = 1e-08, maxiter = 10000)
    x_new <- sol$x
    
    # 平均0に
    x_new <- x_new - mean(x_new)
    
    # stress 計算
    new_diff <- outer(x_new, x_new, FUN = "-")
    norm_diff <- abs(new_diff)
    stress <- 0.5 * sum(W * (norm_diff - dist_matrix)^2)
    
    rel_change <- (stress_prev - stress) / stress_prev
    if (!is.na(rel_change) && rel_change < tol) {
      cat(sprintf("Converged after %d iterations, stress = %.4e\n", iter, stress))
      return(x_new)
    }
    
    x <- x_new
    stress_prev <- stress
  }
  
  cat(sprintf("Max iterations reached, final stress = %.4e\n", stress_prev))
  return(x)
}

# ---------------------------------------------------------
# グラフのノード情報を再取得 (重複削除後)
# ---------------------------------------------------------
nodes_now <- V(g)  # igraphの頂点オブジェクト
N <- length(nodes_now)

# ノードの属性をデータフレーム化
g_nodes_df <- data.frame(
  idx = V(g)$name,    
  ID  = V(g)$ID,
  Gen = V(g)$Gen,
  CV  = V(g)$CV,
  stringsAsFactors = FALSE
)

# Xベクトルを行列としてまとめる
#   重複削除の段階で消えたものはないので再抽出
nodeX_list <- lapply(nodes_now, function(v) {
  unlist(v[x_cols], use.names=FALSE)
})
X_mat <- do.call(rbind, nodeX_list)
# 正規化
X_norm <- t( (t(X_mat) - lower) / diff )

# 距離行列 (ユークリッド)
dist_matrix <- as.matrix(dist(X_norm, method="euclidean"))

# ストレス・メジャライゼーションで1D座標を得る
x_coords <- stress_majorization_1d(dist_matrix)

# ノード順序が混乱しないよう、順番を合わせる
g_nodes_df$x_coord <- x_coords


# ---------------------------------------------------------
# 6. 可視化
#   - x_coord(1D埋め込み) を x軸
#   - CV を y軸
#   - エッジは ggraph または手動で描画可
#   - ここでは簡易的に ggplot + geom_segment でエッジを描いてみる例
# ---------------------------------------------------------
# エッジデータフレーム
edges_now <- as_data_frame(g, what="edges")
colnames(edges_now) <- c("from", "to")

# from/to それぞれの座標を結合
edges_plot_df <- edges_now %>%
  left_join(g_nodes_df, by=c("from"="idx")) %>%
  rename(x_from = x_coord, CV_from = CV) %>%
  left_join(g_nodes_df, by=c("to"="idx")) %>%
  rename(x_to = x_coord, CV_to = CV)

# ノードを sinkかどうか（outdegree=0かどうか）で分類
outdeg <- degree(g, mode="out")
sink_nodes <- names(which(outdeg == 0))
mid_nodes <- setdiff(g_nodes_df$idx, sink_nodes)

# sink かつ CV=0 / CV>0
final_feasible_idx   <- sink_nodes[g_nodes_df$CV[match(sink_nodes, g_nodes_df$idx)] == 0]
final_infeasible_idx <- sink_nodes[g_nodes_df$CV[match(sink_nodes, g_nodes_df$idx)] > 0]

# 中間ノード
mid_feasible_idx   <- mid_nodes[g_nodes_df$CV[match(mid_nodes, g_nodes_df$idx)] == 0]
mid_infeasible_idx <- mid_nodes[g_nodes_df$CV[match(mid_nodes, g_nodes_df$idx)] > 0]

g_nodes_df$category <- "mid_infeasible"
g_nodes_df$category[g_nodes_df$idx %in% mid_feasible_idx]   <- "mid_feasible"
g_nodes_df$category[g_nodes_df$idx %in% final_feasible_idx] <- "final_feasible"
g_nodes_df$category[g_nodes_df$idx %in% final_infeasible_idx] <- "final_infeasible"

# カテゴリごとに色・サイズを設定する例
cat_colors <- c("mid_infeasible"   = "skyblue",
                "mid_feasible"     = "salmon",
                "final_feasible"   = "red",
                "final_infeasible" = "blue")
cat_sizes  <- c("mid_infeasible"   = 1.5,
                "mid_feasible"     = 1.5,
                "final_feasible"   = 3.5,
                "final_infeasible" = 3.5)

# symlog的なスケールにするためには ggplot で pseudo_log_trans などを使用
# (ただし“symlog”を完全に再現するには工夫が要ります)
ggplot() +
  # エッジを矢印付きで描くなら、arrowパラメータを使ってgeom_segment
  geom_segment(
    data = edges_plot_df,
    aes(x = x_from, y = CV_from, xend = x_to, yend = CV_to),
    arrow = arrow(length=unit(0.2,"cm")),
    size = 0.5, alpha=0.3, color="gray"
  ) +
  # ノード散布図
  geom_point(
    data = g_nodes_df,
    aes(x = x_coord, y = CV, color = category, size=category),
    shape = 21, stroke=1
  ) +
  scale_color_manual(values=cat_colors) +
  scale_size_manual(values=cat_sizes) +
  # CV軸を対数スケールっぽく
  scale_y_continuous(trans = scales::pseudo_log_trans(base = 10, sigma=1e-6),
                     limits = c(-1e-6, 1e5)) +
  theme_minimal(base_size=14) +
  labs(x="1D embedding (stress majorization)", y="CV") +
  ggtitle("Constraint Violation vs 1D Embedding")